bgpd/bgp_fsm.c

```diff
@@ -2055,3 +2256,498 @@ int bgp_event_update(struct peer *peer, int event)
+		break;
+	}
+
+	return peer_gr_cmd;
+}
+
+const char *print_global_gr_mode(enum global_mode gl_mode)
+{
+	const char *global_gr_mode = NULL;
+
+	switch (gl_mode) {
+	case GLOBAL_HELPER:
+		global_gr_mode = "GLOBAL_HELPER";
+		break;
+	case GLOBAL_GR:
+		global_gr_mode = "GLOBAL_GR";
+		break;
+	case GLOBAL_DISABLE:
+		global_gr_mode = "GLOBAL_DISABLE";
+		break;
+	case GLOBAL_INVALID:
+		global_gr_mode = "GLOBAL_INVALID";
+		break;
+	}
+
+	return global_gr_mode;
+}
+
+const char *print_global_gr_cmd(enum global_gr_command gl_gr_cmd)
+{
+	const char *global_gr_cmd = NULL;
+
+	switch (gl_gr_cmd) {
+	case GLOBAL_GR_CMD:
+		global_gr_cmd = "GLOBAL_GR_CMD";
+		break;
+	case NO_GLOBAL_GR_CMD:
+		global_gr_cmd = "NO_GLOBAL_GR_CMD";
+		break;
+	case GLOBAL_DISABLE_CMD:
+		global_gr_cmd = "GLOBAL_DISABLE_CMD";
+		break;
+	case NO_GLOBAL_DISABLE_CMD:
+		global_gr_cmd = "NO_GLOBAL_DISABLE_CMD";
+		break;
+	}
+
+	return global_gr_cmd;
+}
+
+enum global_mode bgp_global_gr_mode_get(struct bgp *bgp)
+{
+	return bgp->global_gr_present_state;
+}
+
+enum peer_mode bgp_peer_gr_mode_get(struct peer *peer)
+{
+	return peer->peer_gr_present_state;
+}
+
+int bgp_neighbor_graceful_restart(struct peer *peer,
+						int peer_gr_cmd)
+{
+	enum peer_mode  peer_new_state = PEER_INVALID;
+	enum peer_mode  peer_old_state = PEER_INVALID;
+	struct bgp_peer_gr peer_state;
+	int result = BGP_GR_FAILURE;
+
+	/*
+	 * fetch peer_old_state from peer structure also
+	 * fetch global_old_state from bgp structure,
+	 * peer had a back pointer to bgpo struct ;
+	 */
+
+	if (BGP_DEBUG(graceful_restart, GRACEFUL_RESTART))
+		zlog_debug(
+		"%s [BGP_GR] START:Peer: (%s) : peer_gr_cmd :%s:",
+		  __func__, peer->host, print_peer_gr_cmd(peer_gr_cmd));
+
+	peer_old_state = bgp_peer_gr_mode_get(peer);
+
+	if (peer_old_state == PEER_INVALID) {
+		zlog_debug(
+			"[BGP_GR] peer_old_state == Invalid state !");
+		return BGP_ERR_GR_OPERATION_FAILED;
+	}
+
+	peer_state = peer->PEER_GR_FSM[peer_old_state][peer_gr_cmd];
+	peer_new_state = peer_state.next_state;
+
+	if (peer_new_state == PEER_INVALID) {
+		zlog_debug(
+			"[BGP_GR] Invalid bgp graceful restart command used !");
+		return BGP_ERR_GR_INVALID_CMD;
+	}
+
+	if (peer_new_state != peer_old_state) {
+		result = peer_state.action_fun(peer,
+						peer_old_state,
+						peer_new_state);
+	} else {
+		if (BGP_DEBUG(graceful_restart, GRACEFUL_RESTART))
+			zlog_debug(
+			"[BGP_GR] peer_old_state == peer_new_state !");
+		return BGP_GR_NO_OPERATION;
+	}
+
+	if (result == BGP_GR_SUCCESS) {
+
+		/* Update the mode i.e peer_new_state into the peer structure */
+		peer->peer_gr_present_state = peer_new_state;
+		if (BGP_DEBUG(graceful_restart, GRACEFUL_RESTART))
+			zlog_debug("[BGP_GR] Succesfully change the state of the peer to : %s : !",
+					print_peer_gr_mode(peer_new_state));
+
+		return BGP_GR_SUCCESS;
+	}
+
+	return result;
+}
+
+unsigned int bgp_peer_gr_action(struct peer *peer,
+				int old_peer_state, int new_peer_state)
```

Comment:
can't those states be converted to enum? to clearly print them in debug/error messages?

Reply:
They are already enum. Pls look at enum peer_mode & enum global_mode.